packages/react-scripts/scripts/eject.js

```diff
@@ -36,6 +37,31 @@ prompt(
     process.exit(1);
   }
 
+  // Make sure there are no dirty git status
+  function statusSync() {
+    try {
+      let stdout = execSync(`git status --porcelain`).toString();
+      let status = { dirty: 0, untracked: 0 };
+      stdout.trim().split(/\r?\n/).forEach(file => {
+        if (file.substr(0, 2) === '??') status.untracked++;
+        else status.dirty++;
+      });
+      return status;
+    } catch (e) {
+      return false;
+    }
+  }
+
+  const status = statusSync();
+  if (status && status.dirty) {
+    console.error(
+      `Git repository has ${status.dirty} dirty ${status.dirty > 1 ? 'files' : 'file'}. ` +
+        'We cannot continue as you would lose all the changes in that file or directory. ' +
+        'Please push commit before and run this command again.'
```

Comment:
I think this error message covers all the right bases. The grammar however is a bit confusing.

The first sentence mentions that you have 1 or more dirty files, but then the second sentence only refers to a single file or directory. The third sentence isn't totally clear grammatically and it states that you need to push as well as commit in order to make the directory clean. While pushing may be ideal, it isn't the same as committing and only committing is really necessary here.

What do you think of this as the error message instead?

```javascript
`This git repository has ${status.dirty} dirty ${status.dirty > 1 ? 'files' : 'file'}. ` +
'We cannot continue as you would lose all of your changes. ' +
'Please commit your changes with `git commit` and then run this command again.'
```

This way, the user knows what to do to fix the problem and the message is completely clear.